client/plots/gsea.js

```diff
@@ -0,0 +1,293 @@
+				min: 0,
+				max: 1
+			},
+			{
+				label: 'P-value filter type',
+				type: 'radio',
+				chartType: 'gsea',
+				settingsKey: 'adjusted_original_pvalue',
+				title: 'Toggle between original and adjusted pvalues for volcano plot',
+				options: [
+					{ label: 'adjusted', value: 'adjusted' },
+					{ label: 'original', value: 'original' }
+				]
+			}
+		]
+
+		const geneSet = {
+			label: 'Gene set group',
+			type: 'dropdown',
+			chartType: 'gsea',
+			settingsKey: 'pathway',
+			title: 'Display table showing original and adjusted pvalues corresponding to each significant pathway',
+			boxLabel: '',
+			options: [
+				{ label: 'BP: subset of GO', value: 'BP: subset of GO' },
+				{ label: 'MF: subset of GO', value: 'MF: subset of GO' },
+				{ label: 'CC: subset of GO', value: 'CC: subset of GO' },
+				{ label: 'WikiPathways subset of CP', value: 'WikiPathways subset of CP' },
+				{ label: 'REACTOME subset of CP', value: 'REACTOME subset of CP' },
+				{ label: 'H: hallmark gene sets', value: 'H: hallmark gene sets' }
+			]
+		}
+		if (!this.settings.pathway) {
+			geneSet.options.unshift({ label: '-', value: '-' })
+			this.settings.pathway = '-'
+		}
+		inputs.push(geneSet)
+
+		this.components = {
+			controls: await controlsInit({
+				app: this.app,
+				id: this.id,
+				holder: this.dom.controlsDiv,
+				inputs: inputs
+			})
+		}
+		this.components.controls.on('downloadClick.gsea', () => {
+			downloadTable(this.table_rows, this.table_cols)
+		})
+	}
+	getState(appState) {
+		const config = appState.plots.find(p => p.id === this.id)
+		if (!config) throw `No plot with id='${this.id}' found`
+		return {
+			config
+		}
+	}
+
+	async main() {
+		this.config = structuredClone(this.state.config)
+		this.settings = this.config.settings.gsea
+		await this.setControls()
+		this.dom.header
+			.style('opacity', 0.6)
+			.style('padding-left', '10px')
+			.style('font-size', '0.75em')
+			.text('GENE SET ENRICHMENT ANALYSIS')
+		render_gsea(this)
+	}
+}
+
+async function render_gsea(self) {
+	/*
+m {}
+- gene
+- logfoldchange
+- averagevalue
+- pvalue
+
+add:
+- vo_circle
+	*/
+	if (self.settings.pathway != '-') {
+		self.dom.detailsDiv.selectAll('*').remove()
+		self.config.gsea_params.geneSetGroup = self.settings.pathway
+		const wait = self.dom.detailsDiv.append('div').text('Loading...')
+		const output = await rungsea(self.config.gsea_params)
+		wait.remove()
+		const table_stats = table2col({ holder: self.dom.detailsDiv })
+		const [t1, t2] = table_stats.addRow()
+		t2.style('text-align', 'center').style('font-size', '0.8em').style('opacity', '0.8').text('COUNT')
+		const addStats = [
+			{
+				label: 'Pathways analyzed',
+				values: Object.keys(output).length
+			}
+		]
+
+		for (const dataRow of addStats) {
+			const [td1, td2] = table_stats.addRow()
+			td1.text(dataRow.label)
+			td2.style('text-align', 'end').text(dataRow.values)
+		}
+
+		// Generating the table
+		const gsea_table_cols = [
```

Comment:
because the columns of this table are constant and not variable, move this `const` to the beginning of script will be fine

Reply:
I have moced the const to the top of the script.